### PR TITLE
Implement damping for open-shell references

### PR DESCRIPTION
**Pull request by susilehtola**
## Description
This PR implements damping for open-shell references, which may be very helpful in cases where e.g. ROHF doesn't converge.

## Status
- [x]  Ready to go

### CODE REVIEW

**Comment by susilehtola** [comment]
The PR also implements damping for UKS, because UKS should inherit the UHF damp_update() method.

File: src/lib/libscf_solver/uhf.cc [hunk 1]
```diff
@@ -372,6 +378,27 @@ void UHF::Hx(SharedMatrix x_a, SharedMatrix IFock_a, SharedMatrix Cocc_a,
...
+    }
+    for(int row = 0; row < Db_->rowspi(h); ++row){
+      for(int col = 0; col < Db_->colspi(h); ++col){
+	double Dold = damping_percentage_ * Db_old_->get(h, row, col);
+	double Dnew = (1.0 - damping_percentage_) * Db_->get(h, row, col);
+	Db_->set(h, row, col, Dold+Dnew);
+      }
+    }
+  }
+  Dt_->copy(Da_);
```
Comment (dgasmith) [inline src/lib/libscf_solver/uhf.cc]: Psi4 Matrices have scale and axpy calls, can you switch out the loops for these calls?

```
Da_->scale(1.0 - damping_percentage_)
Da_->axpy(damping_percentage_, Da_old_)
```

**Comment by dgasmith** [comment]
A little worried damping is not being applied correctly. We build JK matrices from C matrices and not the density, the only time the density is used is in the DIIS error vector and for computing the energy.

@jturney @andysim Did I miss something here?

**Comment by susilehtola** [comment]
I wouldn't know about that, since this patch solved my convergence problems on ROHF calculations, e.g. the example in #423 .

Sure, I can switch out the for loops to the versions you suggested; I just adapted the routines from rhf.cc.

**Comment by dgasmith** [comment]
I could be wrong, it just doesn't appear that damping is quite correct in Psi. Certainly nothing you did, ill try to investigate at a later time. This PR looks good to go.

As newer code is introduced it becomes easier to do certain operations, just try to keep new code up to date instead of propagating less concise operations.

**Pull request merged**